Create RestoreSummaryRequest objects in parallel

jeffkl opened the pull request:
When restoring a lot of projects, the creation of RestoreSummaryRequest objects is slow.  By creating them in parallel large restores are 15% faster.

I also had to update the return value of IPreLoadedRestoreRequestProvider.CreateRequests to IReadOnlyCollection since I'm using a ConcurrentBag<T> in the Parallel.ForEach which doesn't support IList<T>

## Bug

Fixes: https://github.com/NuGet/Home/issues/8737
Regression: No
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: I've implemented the usage of a `Parallel.ForEach` loop so that the creation of requests is done on all CPUs.

## Testing/Validation

Tests Added: No  
Reason for not adding tests:  Existing tests should cover this scenario
Validation:  Profiling shows the creation of requests happening on multiple threads now which reduces restore times by up to 20%

nkolev92 commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs:
```diff
@@ -56,10 +57,18 @@ private IReadOnlyList<RestoreSummaryRequest> GetRequestsFromItems(RestoreArgs re
+            };
+
+            // Parallel.Foreach has an optimization for Arrays, so calling .ToArray() is better and adds almost no overhead
+            Parallel.ForEach(dgFile.Restore.ToArray(), parallelOptions, projectNameToRestore =>
```
I have a concern with adding additional scheduled work. 
We blast the threadpool in too many places. 

How does this perform in smaller solutions? Any meaningful improvement there?

nkolev92 replied:
This change also "clashes" with https://github.com/NuGet/NuGet.Client/pull/3086. 
Once that's merged, I'd expect the gains from this to be lower.

jeffkl replied:
This happens before requests are scheduled.  In my profiling, everything happens in parallel except this part of the restore.  The perf gain is much more apparent in large restores, smaller restores will probably not appear to be any faster.

jeffkl replied:
In my testing, I disabled the loading of settings per project and still see about .5 to 1 second improvement for 700 projects.  So I think this is still needed along with #3086

nkolev92 replied:
> This happens before requests are scheduled

That's a good point. :)

jeffkl commented:
/cc @andygerlicher

zivkan commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs:
```diff
@@ -33,14 +34,14 @@ public class DependencyGraphSpecRequestProvider : IPreLoadedRestoreRequestProvid
         }
 
-        public Task<IReadOnlyList<RestoreSummaryRequest>> CreateRequests(RestoreArgs restoreContext)
+        public Task<IReadOnlyCollection<RestoreSummaryRequest>> CreateRequests(RestoreArgs restoreContext)
```
This is a breaking change on a public method on a public class. I find it very, very difficult to imagine why anyone would call this API, but what's the performance impact if you keep the same datatype and copy the results into a list after the `Parallel.ForEach` completes?

jeffkl replied:
Thanks for the concern.  I did more digging and found that `ToList<T>()` would be slow but `ToArray()` is fast and gets rid of the need to change the return type.  I've updated the PR!

jeffkl committed:
Create RestoreSummaryRequest objects in parallel

When restoring a lot of projects, the creation of RestoreSummaryRequest objects is slow.  By creating them in parallel large restores are 15% faster.
